src/uproot/source/fsspec.py

```diff
@@ -38,9 +38,20 @@ def __init__(self, file_path, **options):
         self._fs, self._file_path = fsspec.core.url_to_fs(file_path, **opts)
 
         if self._use_threads:
-            self._executor = concurrent.futures.ThreadPoolExecutor(
-                max_workers=self._num_workers
-            )
+            if self._fs.async_impl:
+                self._executor = uproot.source.futures.LoopExecutor()
+
+                # Bind the loop to the filesystem
+                async def make_fs():
+                    return fsspec.filesystem(
+                        protocol=self._fs.protocol, loop=self._executor.loop
+                    )
+
+                self._fs = self._executor.submit(make_fs).result()
+            else:
+                self._executor = concurrent.futures.ThreadPoolExecutor(
+                    max_workers=self._num_workers
+                )
         else:
             self._executor = uproot.source.futures.TrivialExecutor()
```

Comment:
The decision to use an event loop, a thread pool, or nothing (synchronous/blocking).

I don't see any reason why we should keep the `ThreadPoolExecutor`-based solution around. Maybe there will be some advantage to letting the event loop use multiple threads or each thread has its own event loop, but the main thing that the old `ThreadPoolExecutor`-based solution gets wrong is that it does not send all of the requests immediately.

The switch that falls back on `ThreadPoolExecutor` is `self._fs.async_impl`. If an fsspec backend doesn't implement async, shouldn't we do an event loop anyway? That is, if the function that makes a request makes a request when it is called doesn't return an output until the response is in, couldn't we wrap that up in a future manually?

None of the above should delay this PR, though. It's a next step.

Reply:
If the backend doesn't implement async, then the synchronous interface implies that _some thread_ will block waiting for a request. So if we want any concurrency, we need multiple threads. Event loops will not work any better than a for loop with code that is blocking on IO.

Reply:
I wish I had read this comment before ba7a6bf702e6332fbeb64c6822a29737055601aa 🤦🏻‍♂️ 😄